File: src/main/java/uk/org/stevefisher/swiss/Cmd.java
package uk.org.stevefisher.swiss;

import java.io.BufferedReader;
import java.io.Console;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import uk.org.stevefisher.swiss.logic.BasicSwiss;
import uk.org.stevefisher.swiss.logic.BasicSwiss.Colours;
import uk.org.stevefisher.swiss.logic.PersonScore;
import uk.org.stevefisher.swiss.logic.Player;
import uk.org.stevefisher.swiss.logic.SwissException;

public class Cmd {

	private static String readLine(String prompt) {
		String line = null;
		Console c = System.console();
		if (c != null) {
			line = c.readLine(prompt);
		} else {
			System.out.print(prompt);
			BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
			try {
				line = bufferedReader.readLine();
			} catch (IOException e) {
				// Ignore
			}
		}
		return line;
	}

	public static int getWithDefaults(String name, int defaultValue) {
		String s = readLine(name + " (" + defaultValue + "): ").trim();
		if (s.isEmpty()) {
			return defaultValue;
		} else {
			return Integer.parseInt(s);
		}
	}

	public static void main(String[] args) {
		Cmd cmd = new Cmd();
		cmd.run();
	}

	private void run() {

		int byeScore = getWithDefaults("byeScore", 26);
		int maxCombis = getWithDefaults("maxCombis", 1000000000);
		int enoughGood = getWithDefaults("enoughGood", 100);
		int numLawns = getWithDefaults("numLawns", 2);

		BasicSwiss tournament = new BasicSwiss(byeScore, maxCombis, enoughGood, numLawns);
		Path journal = Paths.get("journal.txt");
		boolean journalRecovery = false;
		List<String> lines = new ArrayList<String>();
		if (Files.exists(journal)) {
			journalRecovery = true;
			System.out.println("Recovering from journal file");
			List<String> names = new ArrayList<String>();
			try (BufferedReader f = new BufferedReader(new FileReader(journal.toFile()))) {
				String[] players = f.readLine().split(",");
				for (String player : players) {

					if (player.toUpperCase().endsWith("-P")) {
						String name = player.substring(0, player.length() - 2).trim();
						names.add(name);
						tournament.addPlayer(name, Colours.PRIMARY, true);
					} else if (player.toUpperCase().endsWith("-S")) {
						String name = player.substring(0, player.length() - 2).trim();
						names.add(name);
						tournament.addPlayer(name, Colours.SECONDARY, true);
					} else {
						String name = player.trim();
						names.add(name);
						tournament.addPlayer(name, true);
					}
				}
				String line;
				while ((line = f.readLine()) != null) {
					lines.add(line);
				}
				tournament.start(names);
			} catch (Exception e) {
				System.err.println(e.getMessage());
				System.exit(1);
			}
		} else {
			while (true) {
				String player = readLine("player (empty to end, finish name with -P or -S if CVD): ");
				if (player.isEmpty()) {
					break;
				}
				try {
					if (player.toUpperCase().endsWith("-P")) {
						tournament.addPlayer(player.substring(0, player.length() - 2).trim(), Colours.PRIMARY);
					} else if (player.toUpperCase().endsWith("-S")) {
						tournament.addPlayer(player.substring(0, player.length() - 2).trim(), Colours.SECONDARY);
					} else {
						tournament.addPlayer(player.trim());
					}
				} catch (SwissException e) {
					System.err.println(e.getMessage());
				}
			}
			try {
				tournament.start();
			} catch (SwissException e) {
				System.err.println(e.getMessage());
				System.exit(1);
			}
		}

		System.out.println("Rounds KO:" + tournament.getKORounds() + ", Max:" + tournament.getMaxRounds()
				+ ", Recommended:" + tournament.getRecRounds());

		int roundNum = 1;
		if (journalRecovery) {
			int roundsInJournal = lines.size() * 2 / tournament.getPlayers().size();
			if ((lines.size() * 2 % tournament.getPlayers().size()) != 0) {
				System.err.println("The journal.txt file has an incomplete round - please correct it manually and try again");
				System.exit(1);
			}
			int linenum = 0;
			for (int roff = 0; roff < roundsInJournal; roff++) {
				List<PersonScore> round = tournament.getRound(roff);
				tournament.makeGamesChoices(round);
				Map<String, Integer> psj = new HashMap<String, Integer>();
				for (int k = 0; k < tournament.getPlayers().size() / 2; k++) {
					String[] bits = lines.get(linenum++).split(",");
					psj.put(bits[0], Integer.parseInt(bits[3]));
					psj.put(bits[2], Integer.parseInt(bits[4]));
				}
				for (PersonScore ps : round) {
					ps.setScore(psj.get(ps.getName()));
				}
				tournament.computeRanking();
				System.out.print("Ranking after round " + roundNum + " ");
				for (String name : tournament.getRanking()) {
					System.out.print(name + ": " + tournament.getPlayers().get(name).getGames() + "  ");
				}
				System.out.println();
				try {
					tournament.prepareRound();
					roundNum++;
				} catch (SwissException e) {
					System.err.println(e.getMessage());
					System.exit(1);
				}
			}
			System.out.println("Recovery of " + roundsInJournal + " rounds complete\n");
		}

		boolean finished = false;
		while (!finished) {
			System.out.println("Starting round " + roundNum);
			List<PersonScore> round = tournament.getRound(roundNum - 1);
			tournament.setByeScores();
			tournament.makeGamesChoices(round);
			boolean roundInProgress = true;
			while (roundInProgress) {
				String cmd = readLine("Game number, S(tatus), E(nd round) ").toUpperCase().trim();
				if ("S".equals(cmd)) {
					printStatus(round);
				} else if ("E".equals(cmd)) {
					if (isExitAllowed(round)) {
						roundInProgress = false;
					}
				} else {
					try {
						int gameNum = Integer.parseInt(cmd);
						storeScores(round, gameNum);
					} catch (NumberFormatException e) {
						System.out.println("Invalid input");
					}
				}
			}

			try {
				tournament.writeLog(round, roundNum == 1);
			} catch (SwissException e) {
				System.out.println("Failed to write log");
			}

			tournament.computeRanking();
			System.out.print("Ranking after round " + roundNum + " ");
			for (String name : tournament.getRanking()) {
				System.out.print(name + ": " + tournament.getPlayers().get(name).getGames() + "  ");
			}
			System.out.println();
			boolean finishChoiceMade = false;
			while (!finishChoiceMade) {
				if (roundNum >= tournament.getRecRounds()) {
					System.out.println("You have completed the recommended number of rounds.");
				}
				if (roundNum >= tournament.getMaxRounds()) {
					System.out.println("You have completed the maximum number of rounds.");
					finished = true;
					finishChoiceMade = true;
					continue;
				}
				String cmd = readLine("FINISH (tournament), NEXT (to start next round) ").toUpperCase().trim();
				if ("FINISH".equals(cmd)) {
					finished = true;
					finishChoiceMade = true;
				} else if ("NEXT".equals(cmd)) {
					try {
						tournament.prepareRound();
					} catch (SwissException e) {
						System.out.println(e.getMessage());
						System.out.println("Tournament will finish");
						finished = true;
						finishChoiceMade = true;
						continue;
					}
					roundNum++;
					finishChoiceMade = true;
				}
			}
		}

		Path source = Paths.get("journal.txt");
		String strDate = new SimpleDateFormat("yyyy-MM-dd").format(Calendar.getInstance().getTime());
		Path target = Paths.get("gamelog-" + strDate + ".txt");
		try {
			Files.move(source, target, StandardCopyOption.REPLACE_EXISTING);
		} catch (IOException e) {
			System.out.println("You have failed to rename " + source + " to " + target);
		}

		Map<String, Integer> fr = tournament.getFinalRanking();
		System.out.println();

		System.out.format("%2s %-18.18s %4.4s : %5.5s %-12.12s %9.9s %6.6s%n", " #", "name", "wins", "hoops", "lawns",
				"primaryXS", "starts");
		for (int i = 1; i <= tournament.getPlayers().size(); i++) {
			for (Entry<String, Integer> entry : fr.entrySet()) {
				if (entry.getValue() == i) {
					String name = entry.getKey();
					int pos = entry.getValue();
					Player p = tournament.getPlayers().get(name);
					int wins = p.getGames();
					int hoops = p.getHoops();
					int prim = p.getPrimaryExcess();
					List<Integer> lawns = new ArrayList<>();
					for (int j = 0; j < numLawns; j++) {
						lawns.add(p.getLawnCount(j));
					}
					int starts = p.getStartCount();

					System.out.format("%2d %-18.18s %4d : %5d %-12.12s %9d %6d%n", pos, name, wins, hoops,
							lawns.toString(), prim, starts);
				}
			}
		}
	}

	private static void printStatus(List<PersonScore> round) {
		int ngame = 1;
		for (int i = 0; i < round.size() / 2; i++) {
			PersonScore p1 = round.get(2 * i);
			PersonScore p2 = round.get(2 * i + 1);
			if (!p1.getName().equals("Bye") && !p2.getName().equals("Bye")) {
				System.out.println("Game " + ngame++ + ": " + p1 + " vs " + p2);
			}
		}
	}

	private static void storeScores(List<PersonScore> round, int gameNum) {
		int ngame = 1;
		for (int i = 0; i < round.size() / 2; i++) {
			PersonScore p1 = round.get(2 * i);
			PersonScore p2 = round.get(2 * i + 1);

			if (p1.getName().equals("Bye") || p2.getName().equals("Bye")) {
				// NOP
			} else if (gameNum == ngame++) {
				if (p1.getScore() != 0 || p2.getScore() != 0) {
					if (!readLine("Enter OVERWRITE to change score ").toUpperCase().trim().equals("OVERWRITE")) {
						return;
					}
				}
				int s1;
				int s2;
				try {
					s1 = Integer.parseInt(readLine("Score for: " + p1.getName() + ": "));
					s2 = Integer.parseInt(readLine("Score for: " + p2.getName() + ": "));
				} catch (NumberFormatException e) {
					System.out.println("Scores must be integers");
					return;
				}
				if (s1 == s2) {
					System.out.println("Draws are not allowed");
				} else {
					p1.setScore(s1);
					p2.setScore(s2);
				}
			}
		}

	}

	private static boolean isExitAllowed(List<PersonScore> round) {
		boolean exitAllowed = true;
		int ngame = 1;
		for (int i = 0; i < round.size() / 2; i++) {
			PersonScore p1 = round.get(2 * i);
			PersonScore p2 = round.get(2 * i + 1);
			if (p1.getName().equals("Bye") || p2.getName().equals("Bye")) {
				// NOP
			} else {
				if (p1.getScore() == 0 && p2.getScore() == 0) {
					System.out.println("Game " + ngame + " has no score recorded");
					exitAllowed = false;
				}
				ngame++;
			}
		}
		return exitAllowed;
	}

}

File: swiss/Cmd.py
from swiss import Logic
import logging
from pathlib import Path
from datetime import datetime

logging.basicConfig(filename='cmd.log', filemode='w', level=logging.INFO)
logger = logging.getLogger(__name__)

class Cmd(object):

    @staticmethod
    def getWithDefaults(name, defaultValue):
        s = input(name + " (" + str(defaultValue) + "): ").strip();
        if s == "": return defaultValue
        if s[0].upper() == "T": return True
        if s[0].upper() == "F": return False
        return int(s)

    def run(self):
        logger.info("We're running")
        tournament = Logic.Tournament()
        journal = Path("journal.txt")
        journalRecovery = journal.is_file()
        lines = []
        if journalRecovery:
            tournament.recoverFromLog(journal)
            tournament.prepareRound()
        
        else: # No journal recovery
            byeScore = Cmd.getWithDefaults("byeScore", 26);
            maxCombis = Cmd.getWithDefaults("maxCombis", 1000000000);
            enoughGood = Cmd.getWithDefaults("enoughGood", 100);
            numLawns = Cmd.getWithDefaults("numLawns", 2);
            randomStart = Cmd.getWithDefaults("randomStart", True);
            tournament.setOpts(byeScore, maxCombis, enoughGood, numLawns, randomStart)
            while True:
                player = input("player (empty to end, finish name with -P or -S if CVD): ").strip()
                if player == "": break
                if player.upper()[-2:] == "-P":
                    tournament.addPlayer(player[:-2].strip(), Logic.Colours.PRIMARY)
                elif player.upper()[-2:] == "-S":
                    tournament.addPlayer(player[:-2].strip(), Logic.Colours.SECONDARY)
                else:
                    tournament.addPlayer(player.strip())
            tournament.start()
          
        print("Rounds KO:" + str(tournament.getKORounds()) + ", Max:" + str(tournament.getMaxRounds())
                + ", Recommended:" + str(tournament.getRecRounds()))

        finished = False
        
        while not finished:
            roundNum = len(tournament.rounds)
            round = tournament.rounds[roundNum - 1]
            print("Starting round", roundNum)
            tournament.setByeScores()
            tournament.makeGamesChoices(round)
            roundInProgress = True
            while roundInProgress:
                cmd = input("Game number, S(tatus), E(nd round) ").upper().strip()
                if cmd == "S":
                    ngame = 1
                    for i in range(len(round) // 2):
                        p1 = round[2 * i]
                        p2 = round[2 * i + 1]
                        if p1[0] != "Bye" and p2[0] != "Bye":
                            print("Game",ngame, ":",p1,"vs",p2)
                            ngame += 1
                elif cmd == "E":
                    if Cmd.isExitAllowed(round): roundInProgress = False        
                else:
                    try:
                        ngame = 1
                        for i in range(len(round) // 2):
                            p1 = round[2 * i]
                            p2 = round[2 * i + 1]
                            if p1[0] != "Bye" and p2[0] != "Bye":
                                ngame += 1
                            if ngame - 1 == int(cmd):
                                Cmd.storeScores(round, i)
                                break
                    except ValueError:
                        print("Invalid input")
             
            try:
                tournament.writeLog(round, roundNum == 1)
            except Exception:
                print("Failed to write log for round", round)

            tournament.computeRanking()
            print("Ranking after round " + str(roundNum) + ":", end="")
            for name in tournament.ranking:
                print(" " + name + ":" + str(tournament.players[name].games), end="")
            print()
            
            finishChoiceMade = False
            while not finishChoiceMade:
                if roundNum >= tournament.getRecRounds(): print("You have completed the recommended number of rounds.")
                if roundNum >= tournament.getMaxRounds():
                    print("You have completed the maximum number of rounds.");
                    finished = True
                    finishChoiceMade = True
                    continue
                
                cmd = input("FINISH (tournament), NEXT (to start next round) ").upper().strip()
                if cmd == "FINISH":
                    finished = True;
                    finishChoiceMade = True;
                elif cmd == "NEXT":
                    try:
                        tournament.prepareRound()
                        finishChoiceMade = True
                    except Exception as e:
                        print(e)
                        print("Tournament will finish")
                        finished = True
                        finishChoiceMade = True
                        continue
  
        source = Path("journal.txt")
        strDate = datetime.today().strftime("%Y-%m-%d")
        target = Path("gamelog-" + strDate + ".txt")
        try:
            source.rename(target)
        except Exception as e:
            print("You have failed to rename", source, "to", target)
            print(e)

        fr = tournament.getFinalRanking();
        print(fr)

        fmt = "{:2} {:.<18} {:4} : {:5} {:>12} {:>10} {:>9}"
        print(fmt.format(" #", "name", "wins", "hoops", "lawns", "primaryXS", "starts"))
        for i in range(1,len(tournament.players)+1):
            for name, val in fr.items():
                if val == i:
                    p = tournament.players[name]
                    wins = p.games
                    hoops = p.hoops
                    prim = p.primarys-p.secondarys
                    lawns = []
                    for j in range(tournament.numLawns):
                        if j in p.lawns: lawns.append(p.lawns[j])
                        else: lawns.append(0)
                    starts = p.startCount
                    print(fmt.format(i, name, wins, hoops, str(lawns), prim, starts))

    @staticmethod
    def storeScores(round, i):
        '''
        Store score at offset i in the round.
        '''
        if i < 0 or i >= len(round) // 2:
            print ("Invalid game number")
            return
        p1 = round[2 * i]
        p2 = round[2 * i + 1]
        
        if p1[0] == "Bye" or p2[0] == "Bye":
            print ("Invalid game number")
            return
           
        if p1[1] != 0 or p2[1] != 0:
            if input("Enter OVERWRITE to change score ").upper().strip() != "OVERWRITE": return
    
        try:
            s1 = int(input("Score for: " + p1[0] + ": "))
            s2 = int(input("Score for: " + p2[0] + ": "))
        except ValueError:
            print("Scores must be integers")
            return
        
        if s1 == s2:
            print("Draws are not allowed")
            return
        
        p1[1] = s1
        p2[1] = s2

    @staticmethod
    def isExitAllowed(round):
        exitAllowed = True
        ngame = 1;
        for i in range(len(round) // 2):
            p1 = round[2 * i]
            p2 = round[2 * i + 1]
            if p1[0] != "Bye" and p2[0] != "Bye":
                if p1[1] == 0 and p2[1] == 0:
                    print("Game", ngame, "has no score recorded")
                    exitAllowed = False
                ngame += 1
        return exitAllowed

def main():
    cmd = Cmd()
    cmd.run()

if __name__ == "__main__":
    main()
